src/environmentd/tests/sql.rs

```diff
@@ -1863,7 +1863,16 @@ async fn test_session_linearizability(isolation_level: &str) {
         .await
         .unwrap();
 
-    let source_ts = test_util::get_explain_timestamp(pg_table_name, &mz_client).await;
+    // Wait until the Coordinator has learned about at least one upper
+    // advancement in the source, which results in picking a higher timestamp
+    // for queries.
+    let mut source_ts;
+    loop {
+        source_ts = test_util::get_explain_timestamp(pg_table_name, &mz_client).await;
+        if source_ts > 0 {
+            break;
+        }
+    }
```

Comment:
It's still not really clear to me why the query in `test_util::wait_for_pg_table_population` would choose a timestamp at a value larger than the source upper, causing it to return before the upper passes 0.

Reply:
While the upper is still `[0]` (as far as the coord knows), it also holds a since at `[0]`. And the `wait_for_pg_table_population` query will use an as_of of `0`, which is fine and correct. And the data happens to be there in the source shard and the query succeeds, because the real upper in the source shard is already a bit further ahead.

Reply:
Spoke about this in person, the key detail is that the first batch of data in the source will get written at timestamp 0. The next batch will get written at the current system time. This seems not correct, but explains the observed behavior. We've had multiple discussions about writing source data in the past, so I won't rehash that here.